Case insensitive HTTPHeaders, HTTPResponse context manager and some fixes

michalpokusa opened the pull request:
Added separate class for HTTP headers, that allows accessing headers using case-insensitive names.

Also, changed some docstrings the were unclear and one that was misleading as it was inconsistent with the code.

Changed the order of Docs References, so that the most "important" ones are on the top, with enums and dataclass-like classes below.

Updated library version to __1.1.0__ in sphinx configuration file.

Fixes #26
Fixes #28

anecdata commented on adafruit_httpserver/response.py:
```diff
@@ -64,21 +67,20 @@ def _construct_response_bytes(  # pylint: disable=too-many-arguments
-        response_headers.setdefault("Connection", "close")
+        headers.setdefault("Content-Type", content_type)
+        headers.setdefault(
+            "Content-Length", content_length or len(body.encode("utf-8"))
```
Can we avoid encoding the body twice?

michalpokusa replied:
Nice spot, it is the little things that matter.
Thanks!

Changed in [this commit](https://github.com/adafruit/Adafruit_CircuitPython_HTTPServer/pull/29/commits/aeb95a293251192b4ab3cea61a8fb70d9cee4d51).

anecdata commented:
@michalpokusa Any thoughts on this 404 issue? <https://github.com/adafruit/Adafruit_CircuitPython_HTTPServer/issues/28#issuecomment-1362933127> I don't see anything obvious in this PR that would have affected it.

michalpokusa commented:
> @michalpokusa Any thoughts on this 404 issue? [#28 (comment)](https://github.com/adafruit/Adafruit_CircuitPython_HTTPServer/issues/28#issuecomment-1362933127) I don't see anything obvious in this PR that would have affected it. Maybe it came in with the refactor and just wasn't noticed until now. Worth a separate issue?

Working on it, also I see that other PR was merged and there are merge conflicts, so I will fix both before proceeding

michalpokusa commented:
After merging recently updated `main` branch i noticed some inconsistencies wit hthe `HTTPResponse` API, mainy the fact that some function would accept `conn` and some not. Also in chunked response sending headers and empty chunk to finish transfer should be implicit.

I decided to modify the `HTTPResposne` API to be unified by using context managers, by doings so i was able to reduce abstraction and number of lines of actual code, the PR may not show this as I added/extended some docstrings. 

Also fixed the 404 problem in #28.

paul-1 commented:
That for sure looks a lot cleaner.  Thanks for doing that

michalpokusa commented:
In the meantime I updated examples to new "context manager way" of serving response. Did some testing on my own and I believe that it works as it should. Also I added raising error if someone tries to use `.send()` or `.send_file()` multiple times per response.

paul-1 commented on adafruit_httpserver/response.py:
```diff
@@ -14,184 +14,206 @@
-    def send_body_chunk(
-        self, conn: Union["SocketPool.Socket", "socket.socket"], chunk: str
+    def send_file(
+        self,
+        filename: str = "index.html",
+        root_path: str = "./",
     ) -> None:
-        """Send chunk of data to the given socket.  Send an empty("") chunk to finish the session.
+        """
+        Send response with content of ``filename`` located in ``root_path``.
+        Implicitly calls ``_send_headers`` before sending the file content.
+
+        Should be called **only once** per response.
+        """
+        if self._response_already_sent:
+            raise RuntimeError("Response was already sent")
+
+        if not root_path.endswith("/"):
+            root_path += "/"
```
This maybe just me, but this creates different usages for the "/" between the route and the send_file command,
```
@server.route("/banner.png")
def base(request):
    hdr = HTTPHeaders({"Cache-Control": "max-age=31536000"})
    with HTTPResponse(request, headers=hdr) as res:
        res.send_file(root_path='/wwwroot/', filename='banner.png')
```
Note: the route needs the leading slash on the route file name to match, but when sending the file, the slash is at the end of the root_path, and not in the filename.

michalpokusa replied:
Could you elaborate on that?
The `route` leading "/" is similar to Flask, that is it is a pathname similar to JS's `location.pathname`.
![image](https://user-images.githubusercontent.com/72110769/209758562-b86462c7-db8b-4a87-bc61-9a6f467daa8e.png)

The `send_file` leading and trailing "/" is optional, in your example all `wwwroot`, `/wwwroot` and `/wwwroot/` would result in the same, that is serving the file e.g. `H:\wwwroot\banner.png`. The only way I see it could be changed is by using `os.chdir()` somewhere in the code, that could mess up the relative paths.

Another thing is that in your example it might be better to use `root_path` in `HTTPServer.start()` for serving static files, thus eliminating the need for a `route` and handler.

Overall the `HTTPServer.start(root_path=...)` determines the directory where the server will look for files if **no handler is found**, so if a url is `http://192.168.0.2/wwwroot/banner.png` and there is not handler it wil look for `/<server_root_path>/wwwroot/banner.png` file in file system.

The `root_path` in `send_file` has the same purpose, but is not connected to server's `root_path`, that enables returning only selected files from non-shared directory, as in your example above e.g. '/wwwroot/style.css' is not accessible without a handler.

And finally the `filename` is simply the file's name, irrelevant to it's location in file system.

paul-1 replied:
Yes, I have the default root_path set in the server start line, and that does work.   However, I have to define a route handler for each of my files, since I want to add custom headers to the files.
```HTTPHeaders({"Cache-Control": "max-age=604800"})```

Anyway, I must have just changed too many things at once.

...as this works
```
@server.route("/css/wire.css")
def base(request):
    hdr = HTTPHeaders({"Cache-Control": "max-age=604800"})
    with HTTPResponse(request, headers=hdr) as res:
        res.send_file(root_path='/wwwroot/', filename='css/wire.css')
```
and so does this.
```
@server.route("/css/wire.css")
def base(request):
    hdr = HTTPHeaders({"Cache-Control": "max-age=604800"})
    with HTTPResponse(request, headers=hdr) as res:
        res.send_file(root_path='/wwwroot', filename='/css/wire.css')
```
the file operations do not care if there is a '//' in the path.

michalpokusa replied:
Thanks for elaborating. Does this mean that it is working as expected fro you?

Also I belive it should be possible to reuse the same handler with many `route` decorators and then set `filename=request.path`, eliminating the need for multiple handlers and hard coding returned file name.

paul-1 replied:
Yes, everything is working as expected.

Thanks for the code savings tip.  That works fine too.

paul-1 reviewed (commented):
I tested my use cases, and everything seemed to work fine, with exception of sending html in chunked mode.

michalpokusa commented on adafruit_httpserver/response.py:
```diff
@@ -14,184 +14,206 @@
-                self._send_bytes(conn, bytes_read)
+    def __enter__(self):
+        if self.chunked:
+            self._send_headers()
```
That was a bug indeed. Fixed that in [d14a13a](https://github.com/adafruit/Adafruit_CircuitPython_HTTPServer/pull/29/commits/d14a13a1c472f7b9aa8d6d3519732012bd62a735). Now it should be working fine.

paul-1 replied:
Yes, works fine now.

michalpokusa commented:
@anecdata Could be please take a look at the current state of this PR. I believe there is everything I wanted to add and that it is ready to be merged.

In the [c758e51 commit](https://github.com/adafruit/Adafruit_CircuitPython_HTTPServer/pull/29/commits/c758e519c0bb266e191d955e28706e48791c1d82) I also "included" the #30.

anecdata reviewed (commented):
Tested file serving, GET, POST, a couple of MIME types, custom HTTP status code, chunked and non-chunked response bodies.

This is a pretty large change in API, so we should change the version number accordingly.

This looks good to me, one minor thing.

michalpokusa commented on adafruit_httpserver/request.py:
```diff
@@ -8,17 +8,36 @@
+    Socket object usable to send and receive data on the connection.
+    """
+
+    address: Tuple[str, int]
```
I agree. I changed it to `client_address`. When it comes to the version, previously I changed it to 1.1.0 as the "context-manager" way of serving responses is indeed a big change, do you think it is ok or should it be 2.0.0?

anecdata replied:
1.1.0 was PR 22, not sure why that does't show up in the Releases sidebar as the latest version. I think it gets assigned at release time, probably 1.2.0 is sufficient if that's enough of a clue that people need to rewrite their routes.

michalpokusa replied:
You're right, that is where I checked when I changed it for the first time. Changing it to 1.2.0 then.

anecdata replied:
I'm not 100% sure, but I don't think you have to change that file, the version gets assigned semi-automatically when the merge is released. In the `.py` file in this repo, [version](https://github.com/adafruit/Adafruit_CircuitPython_HTTPServer/blob/5de66faf9a4ac271f07b23c21e9910d0a7e7fa16/adafruit_httpserver/__init__.py#L22) is:
`__version__ = "0.0.0+auto.0"`.

michalpokusa replied:
Ok, so, I'm changing it back to the "1.0" as i see other libs leave it unchanged. 
[adafruit_requests](https://github.com/adafruit/Adafruit_CircuitPython_Requests/blob/main/docs/conf.py#L62)
[adafruit_minimqtt](https://github.com/adafruit/Adafruit_CircuitPython_MiniMQTT/blob/main/docs/conf.py#L61)

anecdata commented:
1.1.0 was PR 22, not sure why that does't show up in the Releases sidebar as the latest version. I think it gets assigned at release time, probably 1.2.0 is sufficient if that's enough of a clue that people need to rewrite their routes.

anecdata reviewed (approved):
Thanks, @michalpokusa. And thanks @bill88t for #30 included here.

dhalbert commented:
This should be 2.0.0 if there are incompatible API or use changes. A major version change means a code rewrite is needed.

michalpokusa commented:
> This should be 2.0.0 if there are incompatible API or use changes. A major version change means a code rewrite is needed.

If I understand correctly, the code stays the same, only tag changes, so no futher edits are needed on my part to make it 2.0.0

bill88t commented:
Closing #30 then. Thanks for taking care of it!

dhalbert commented:
> If I understand correctly, the code stays the same, only tag changes, so no further edits are needed on my part to make it 2.0.0

Right, when I or whoever makes a release, it will be 2.0.0. There _are_ incompatible API or use changes, right?

michalpokusa commented:
>There _are_ incompatible API or use changes, right?

Yes, instead of returning HTTPResponse instance in route handlers, the function should call a method on HTTPResponse, so all handlers need to be slightly modified, but still.

dhalbert commented:
Great - I will make a release now.

I only found one use of this library in the Learn guides:
https://github.com/adafruit/Adafruit_Learning_System_Guides/blob/main/PicoW_CircuitPython_HTTP_Server/code.py
Could someone take a look at that and see if there are any changes to be made? Thanks.

michalpokusa commented:
> Great - I will make a release now.
> 
> I only found one use of this library in the Learn guides: https://github.com/adafruit/Adafruit_Learning_System_Guides/blob/main/PicoW_CircuitPython_HTTP_Server/code.py Could someone take a look at that and see if there are any changes to be made? Thanks.

Thanks for merging.

Yes, there will be a small change in `base` and `buttonpress` functions. I could make the required changes, but I will be unable to test it as I do not own a Raspberry Pico W.

anecdata commented:
Maybe if Liz still has it set up, it would be easy for her to test. I have a Pico W and could test the routes, but I don't have the peripherals, so I'd stub out the I/O (possibly introducing or masking an issue).

Any network code  written for `raspberrypi` _should_ run on `espressif` (but not necessarily vice versa since there are some things not implemented / not implementable).

dhalbert commented:
If you either of you could make a draft PR, I'll test it by simplifying or setting up the base hardware. Thanks.

anecdata commented:
https://github.com/adafruit/Adafruit_Learning_System_Guides/pull/2375
